File: app/agents/query_router.py
from __future__ import annotations

from pathlib import Path

from app.schemas.document import Manifest
from app.schemas.retrieval import RouteDecision
from app.text_utils import score_text, slugify, tokenize


class QueryRouter:
    def __init__(self, knowledge_base_dir: Path, manifest: Manifest):
        self.knowledge_base_dir = knowledge_base_dir
        self.manifest = manifest

    def route(self, query: str, retry_level: int = 1) -> RouteDecision:
        query_terms = set(tokenize(query))
        candidates: dict[str, float] = {}
        for folder in self._indexed_dirs():
            text = self._read_index_text(folder)
            scores = [score_text(query_terms, text)]
            for doc in self.manifest.documents:
                if folder in doc.all_categories or Path(doc.path).parts[:1] == (folder,):
                    scores.append(score_text(query_terms, " ".join([doc.title, doc.summary, *doc.tags])))
            candidates[folder] = max(scores or [0.0])

        ranked = sorted(candidates.items(), key=lambda item: item[1], reverse=True)
        if retry_level >= 4:
            dirs = [name for name, _ in ranked] or self._fallback_dirs()
            reason = "Full indexed-folder search because retry level requests a broader scope."
        else:
            limit = 2 if retry_level <= 1 else 4
            dirs = [name for name, score in ranked if score > 0][:limit]
            if not dirs:
                dirs = [name for name, _ in ranked[:limit]] or self._fallback_dirs()
            reason = "Selected folders from the root index by matching index text and manifest metadata."
        return RouteDecision(target_dirs=dirs, reason=reason)

    def _read_index_text(self, folder: str) -> str:
        chunks = [folder]
        local_index = self.knowledge_base_dir / folder / "local_index.md"
        root_section = self._root_index_section(folder)
        if root_section:
            chunks.append(root_section)
        if local_index.exists():
            chunks.append(local_index.read_text(encoding="utf-8", errors="replace"))
        return "\n".join(chunks)

    def _indexed_dirs(self) -> list[str]:
        root_index = self.knowledge_base_dir / "index.md"
        if not root_index.exists():
            return self._fallback_dirs()
        dirs = []
        for line in root_index.read_text(encoding="utf-8", errors="replace").splitlines():
            stripped = line.strip()
            if stripped.startswith("## ") and not stripped.startswith("### "):
                name = slugify(stripped.lstrip("#").strip())
                if name:
                    dirs.append(name)
        return list(dict.fromkeys(dirs)) or self._fallback_dirs()

    def _root_index_section(self, folder: str) -> str:
        root_index = self.knowledge_base_dir / "index.md"
        if not root_index.exists():
            return ""
        lines = root_index.read_text(encoding="utf-8", errors="replace").splitlines()
        section: list[str] = []
        in_section = False
        for line in lines:
            stripped = line.strip()
            if stripped.startswith("## ") and not stripped.startswith("### "):
                heading = slugify(stripped.lstrip("#").strip())
                if in_section:
                    break
                in_section = heading == folder
            if in_section:
                section.append(line)
        return "\n".join(section)

    def _fallback_dirs(self) -> list[str]:
        if not self.knowledge_base_dir.exists():
            return []
        return sorted(path.name for path in self.knowledge_base_dir.iterdir() if path.is_dir())

File: src/agents/queryRouter.ts
import fs from "node:fs";
import path from "node:path";
import { type PiAgentClient } from "../piAgentClient.js";
import { allCategories, type Manifest, type RouteDecision } from "../types.js";
import { scoreText, slugify, tokenize } from "../textUtils.js";

export class QueryRouter {
  constructor(private knowledgeBaseDir: string, private manifest: Manifest, private agent?: PiAgentClient) {}

  async route(query: string, retryLevel = 1): Promise<RouteDecision> {
    if (this.agent?.available) {
      const decision = await this.piRoute(query, retryLevel);
      if (decision.targetDirs.length) return decision;
    }
    return this.heuristicRoute(query, retryLevel);
  }

  private heuristicRoute(query: string, retryLevel: number): RouteDecision {
    const queryTerms = new Set(tokenize(query));
    const candidates = this.indexedDirs().map((folder) => {
      const text = this.readIndexText(folder);
      const scores = [scoreText(queryTerms, text)];
      for (const doc of this.manifest.documents) {
        if (allCategories(doc).has(folder) || doc.path.split(/[\\/]/)[0] === folder) {
          scores.push(scoreText(queryTerms, [doc.title, doc.summary, ...doc.tags].join(" ")));
        }
      }
      return [folder, Math.max(...scores)] as const;
    });
    const ranked = candidates.sort((a, b) => b[1] - a[1]);
    if (retryLevel >= 4) {
      return { targetDirs: ranked.map(([name]) => name), reason: "Full indexed-folder search because retry level requests a broader scope." };
    }
    const limit = retryLevel <= 1 ? 2 : 4;
    const dirs = ranked.filter(([, score]) => score > 0).slice(0, limit).map(([name]) => name);
    return {
      targetDirs: dirs.length ? dirs : ranked.slice(0, limit).map(([name]) => name),
      reason: "Selected folders from the root index by matching index text and manifest metadata."
    };
  }

  private async piRoute(query: string, retryLevel: number): Promise<RouteDecision> {
    const availableFolders = this.indexedDirs();
    const result = await this.agent?.completeJson<{ targetDirs?: string[]; target_dirs?: string[]; reason?: string }>(
      "Route a user question to knowledge-base folders. Return only JSON with targetDirs and reason. targetDirs must only contain names from availableFolders.",
      JSON.stringify({
        query,
        retryLevel,
        availableFolders,
        rootIndex: safeRead(path.join(this.knowledgeBaseDir, "index.md")).slice(0, 8000)
      })
    );
    const rawDirs = result?.targetDirs ?? result?.target_dirs ?? [];
    const allowed = new Set(availableFolders);
    const dirs = rawDirs.map(String).map(slugify).filter((dir) => allowed.has(dir));
    const limit = retryLevel >= 4 ? availableFolders.length : retryLevel <= 1 ? 2 : 4;
    return { targetDirs: dirs.slice(0, limit), reason: result?.reason ?? "Selected folders with Pi SDK router." };
  }

  private indexedDirs(): string[] {
    const indexPath = path.join(this.knowledgeBaseDir, "index.md");
    if (!fs.existsSync(indexPath)) return fallbackDirs(this.knowledgeBaseDir);
    const dirs = safeRead(indexPath)
      .split(/\r?\n/)
      .map((line) => line.trim())
      .filter((line) => line.startsWith("## ") && !line.startsWith("### "))
      .map((line) => slugify(line.replace(/^#+/, "").trim()));
    return [...new Set(dirs)].filter(Boolean);
  }

  private readIndexText(folder: string): string {
    return [folder, rootIndexSection(this.knowledgeBaseDir, folder), safeRead(path.join(this.knowledgeBaseDir, folder, "local_index.md"))].join("\n");
  }
}

function rootIndexSection(knowledgeBaseDir: string, folder: string): string {
  const lines = safeRead(path.join(knowledgeBaseDir, "index.md")).split(/\r?\n/);
  const out: string[] = [];
  let active = false;
  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed.startsWith("## ") && !trimmed.startsWith("### ")) {
      if (active) break;
      active = slugify(trimmed.replace(/^#+/, "").trim()) === folder;
    }
    if (active) out.push(line);
  }
  return out.join("\n");
}

function fallbackDirs(dir: string): string[] {
  try {
    return fs.readdirSync(dir, { withFileTypes: true }).filter((entry) => entry.isDirectory()).map((entry) => entry.name).sort();
  } catch {
    return [];
  }
}

function safeRead(file: string): string {
  try {
    return fs.readFileSync(file, "utf8");
  } catch {
    return "";
  }
}
